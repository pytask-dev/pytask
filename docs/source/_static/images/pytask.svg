<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 24.1.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Capa_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 438.9 438.9" style="enable-background:new 0 0 438.9 438.9;" xml:space="preserve">
<style type="text/css">
	.st0{fill:#C19C00;}
	.st1{fill:#13A10E;}
	.st2{fill:#C50F1F;}
</style>
<path d="M348,57.5h-39.7V39.7c0-5.7-6.3-8.4-12-8.4h-30.8C258.1,10.5,239.8,0,218.9,0c-20.7-0.8-39.5,11.9-46.5,31.3h-30.3
	c-5.7,0-11.5,2.6-11.5,8.4v17.8H90.9C67.4,57.8,48.1,76.3,47,99.8v299.4c0,23,20.9,39.7,43.9,39.7h257c23,0,43.9-16.7,43.9-39.7
	V99.8C390.7,76.3,371.5,57.8,348,57.5z M151.5,52.3h28.7c5-0.6,9-4.4,9.9-9.4c3.1-13.5,14.9-23.1,28.7-23.5
	c13.7,0.4,25.3,10.1,28.2,23.5c0.9,5.1,5.2,9,10.4,9.4h29.8v41.8H151.5V52.3z M371,399.2c0,11.5-11.5,18.8-23,18.8h-257
	c-11.5,0-23-7.3-23-18.8V99.8c1.1-12,11-21.2,23-21.4h39.7V105c0.6,5.9,5.6,10.2,11.5,9.9h154.1c6,0.3,11.2-4,12-9.9V78.4H348
	c12,0.2,21.9,9.5,23,21.4L371,399.2L371,399.2z"/>
<path class="st0" d="M179.2,233.6c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,250.3c-3.9-4.1-10.4-4.4-14.6-0.5
	c-4,4.2-4,10.9,0,15.2l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14
	C179.6,234,179.4,233.8,179.2,233.6z"/>
<path d="M329.2,256H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,256,329.2,256z"/>
<path class="st1" d="M179.2,150c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,166.7c-3.9-4.1-10.4-4.4-14.6-0.5c-4,4.2-4,10.9,0,15.2
	l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14C179.6,150.4,179.4,150.2,179.2,150z"/>
<path d="M329.2,172.4H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,172.4,329.2,172.4z"
	/>
<path class="st2" d="M179.2,317.2c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,333.9c-3.9-4.1-10.4-4.4-14.6-0.5
	c-4,4.2-4,10.9,0,15.2l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14
	C179.6,317.6,179.4,317.4,179.2,317.2z"/>
<path d="M329.2,339.6H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,339.6,329.2,339.6z"
	/>
</svg>
