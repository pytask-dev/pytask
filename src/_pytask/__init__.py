"""This module should not contain any imports except for the version."""
from __future__ import annotations


try:
    from ._version import version as __version__
except ImportError:  # pragma: no cover
    # broken installation, we don't even try unknown only works because we do poor mans
    # version compare
    __version__ = "unknown"


__all__ = ["__version__"]
