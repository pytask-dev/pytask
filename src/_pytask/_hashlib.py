from __future__ import annotations

import hashlib
import sys
from contextlib import suppress
from pathlib import Path
from typing import Any


if sys.version_info >= (3, 11):  # pragma: no cover
    from hashlib import file_digest
else:  # pragma: no cover
    # This tuple and __get_builtin_constructor() must be modified if a new
    # always available algorithm is added.
    __always_supported = (
        "md5",
        "sha1",
        "sha224",
        "sha256",
        "sha384",
        "sha512",
        "blake2b",
        "blake2s",
        "sha3_224",
        "sha3_256",
        "sha3_384",
        "sha3_512",
        "shake_128",
        "shake_256",
    )

    algorithms_guaranteed = set(__always_supported)
    algorithms_available = set(__always_supported)

    __all__ = __always_supported + (
        "new",
        "algorithms_guaranteed",
        "algorithms_available",
        "file_digest",
    )

    __builtin_constructor_cache = {}

    # Prefer our blake2 implementation
    # OpenSSL 1.1.0 comes with a limited implementation of blake2b/s. The OpenSSL
    # implementations neither support keyed blake2 (blake2 MAC) nor advanced
    # features like salt, personalization, or tree hashing. OpenSSL hash-only
    # variants are available as 'blake2b512' and 'blake2s256', though.
    __block_openssl_constructor = {
        "blake2b",
        "blake2s",
    }

    def __get_builtin_constructor(name):
        cache = __builtin_constructor_cache
        constructor = cache.get(name)
        if constructor is not None:
            return constructor

        with suppress(ImportError):
            if name in {"SHA1", "sha1"}:
                import _sha1

                cache["SHA1"] = cache["sha1"] = _sha1.sha1
            elif name in {"MD5", "md5"}:
                import _md5

                cache["MD5"] = cache["md5"] = _md5.md5
            elif name in {"SHA256", "sha256", "SHA224", "sha224"}:
                try:
                    import _sha2
                except ImportError:
                    import _sha256 as _sha2
                cache["SHA224"] = cache["sha224"] = _sha2.sha224
                cache["SHA256"] = cache["sha256"] = _sha2.sha256
            elif name in {"SHA512", "sha512", "SHA384", "sha384"}:
                try:
                    import _sha2
                except ImportError:
                    import _sha256 as _sha2
                cache["SHA384"] = cache["sha384"] = _sha2.sha384
                cache["SHA512"] = cache["sha512"] = _sha2.sha512
            elif name in {"blake2b", "blake2s"}:
                import _blake2

                cache["blake2b"] = _blake2.blake2b
                cache["blake2s"] = _blake2.blake2s
            elif name in {"sha3_224", "sha3_256", "sha3_384", "sha3_512"}:
                import _sha3

                cache["sha3_224"] = _sha3.sha3_224
                cache["sha3_256"] = _sha3.sha3_256
                cache["sha3_384"] = _sha3.sha3_384
                cache["sha3_512"] = _sha3.sha3_512
            elif name in {"shake_128", "shake_256"}:
                import _sha3

                cache["shake_128"] = _sha3.shake_128
                cache["shake_256"] = _sha3.shake_256

        constructor = cache.get(name)
        if constructor is not None:
            return constructor

        raise ValueError("unsupported hash type " + name)

    def __get_openssl_constructor(name):
        if name in __block_openssl_constructor:
            # Prefer our builtin blake2 implementation.
            return __get_builtin_constructor(name)
        try:
            # MD5, SHA1, and SHA2 are in all supported OpenSSL versions
            # SHA3/shake are available in OpenSSL 1.1.1+
            f = getattr(_hashlib, "openssl_" + name)
            # Allow the C module to raise ValueError.  The function will be
            # defined but the hash not actually available.  Don't fall back to
            # builtin if the current security policy blocks a digest, bpo#40695.
            f(usedforsecurity=False)
            # Use the C function directly (very fast)
            return f
        except (AttributeError, ValueError):
            return __get_builtin_constructor(name)

    def __py_new(name, data=b"", **kwargs):
        """new(name, data=b'', **kwargs) - Return a new hashing object using the
        named algorithm; optionally initialized with data (which must be
        a bytes-like object).
        """
        return __get_builtin_constructor(name)(data, **kwargs)

    def __hash_new(name, data=b"", **kwargs):
        """new(name, data=b'') - Return a new hashing object using the named algorithm;
        optionally initialized with data (which must be a bytes-like object).
        """
        if name in __block_openssl_constructor:
            # Prefer our builtin blake2 implementation.
            return __get_builtin_constructor(name)(data, **kwargs)
        try:
            return _hashlib.new(name, data, **kwargs)
        except ValueError:
            # If the _hashlib module (OpenSSL) doesn't support the named
            # hash, try using our builtin implementations.
            # This allows for SHA224/256 and SHA384/512 support even though
            # the OpenSSL library prior to 0.9.8 doesn't provide them.
            return __get_builtin_constructor(name)(data)

    try:
        import _hashlib

        new = __hash_new
        __get_hash = __get_openssl_constructor
        algorithms_available = algorithms_available.union(
            _hashlib.openssl_md_meth_names
        )
    except ImportError:
        _hashlib = None
        new = __py_new
        __get_hash = __get_builtin_constructor

    with suppress(ImportError):
        # OpenSSL's PKCS5_PBKDF2_HMAC requires OpenSSL 1.0+ with HMAC and SHA
        from _hashlib import pbkdf2_hmac

        __all__ += ("pbkdf2_hmac",)

    with suppress(ImportError):
        # OpenSSL's scrypt requires OpenSSL 1.1+
        from _hashlib import scrypt

    def file_digest(fileobj, digest, /, *, _bufsize=2**18):
        """Hash the contents of a file-like object. Returns a digest object.

        *fileobj* must be a file-like object opened for reading in binary mode.
        It accepts file objects from open(), io.BytesIO(), and SocketIO objects.
        The function may bypass Python's I/O and use the file descriptor *fileno*
        directly.

        *digest* must either be a hash algorithm name as a *str*, a hash
        constructor, or a callable that returns a hash object.
        """
        # On Linux we could use AF_ALG sockets and sendfile() to archive zero-copy
        # hashing with hardware acceleration.
        if isinstance(digest, str):
            digestobj = new(digest)
        else:
            digestobj = digest()

        if hasattr(fileobj, "getbuffer"):
            # io.BytesIO object, use zero-copy buffer
            digestobj.update(fileobj.getbuffer())
            return digestobj

        # Only binary files implement readinto().
        if not (
            hasattr(fileobj, "readinto")
            and hasattr(fileobj, "readable")
            and fileobj.readable()
        ):
            raise ValueError(
                f"'{fileobj!r}' is not a file-like object in binary reading mode."
            )

        # binary file, socket.SocketIO object
        # Note: socket I/O uses different syscalls than file I/O.
        buf = bytearray(_bufsize)  # Reusable buffer to reduce allocations.
        view = memoryview(buf)
        while True:
            size = fileobj.readinto(buf)
            if size == 0:
                break  # EOF
            digestobj.update(view[:size])

        return digestobj


def hash_value(value: Any) -> int | str:
    """Hash values.

    Compute the hash of paths, strings, and bytes with a hash function or otherwise the
    hashes are salted.

    The hash of None constant https://github.com/python/cpython/pull/99541 starting with
    Python 3.12.

    """
    if value is None:
        return 0xFCA86420
    if isinstance(value, (tuple, list)):
        value = "".join(str(hash_value(i)) for i in value)
    if isinstance(value, Path):
        value = str(value)
    if isinstance(value, str):
        value = value.encode()
    if isinstance(value, bytes):
        return str(hashlib.sha256(value).hexdigest())
    return hash(value)
