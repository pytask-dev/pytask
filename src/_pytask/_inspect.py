from __future__ import annotations

__all__ = ["get_annotations"]


from inspect import get_annotations
