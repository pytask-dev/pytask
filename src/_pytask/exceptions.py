"""This module contains custom exceptions."""
from __future__ import annotations


class PytaskError(Exception):
    """Base exception for pytask which should be inherited by all other exceptions."""


class NodeNotFoundError(PytaskError):
    """Exception for missing dependencies."""


class NodeNotCollectedError(PytaskError):
    """Exception for nodes which could not be collected."""


class ConfigurationError(PytaskError):
    """Exception during the configuration."""


class CollectionError(PytaskError):
    """Exception during collection."""


class ResolvingDependenciesError(PytaskError):
    """Exception during resolving dependencies."""


class ExecutionError(PytaskError):
    """Exception during execution."""
