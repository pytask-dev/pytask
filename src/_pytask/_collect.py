"""This module contains helpers for :mod:`_pytask.collect`."""
from __future__ import annotations

import itertools
import uuid
from pathlib import Path
from typing import Any
from typing import Iterable

import attr
from _pytask.exceptions import NodeNotCollectedError
from _pytask.nodes import find_duplicates
from _pytask.nodes import MetaNode
from _pytask.session import Session


__all__ = ["convert_objects_to_node_dictionary"]


def convert_objects_to_node_dictionary(objects: Any, when: str) -> dict[Any, Any]:
    """Convert objects to node dictionary."""
    list_of_dicts = [_convert_to_dict(x) for x in objects]
    _check_that_names_are_not_used_multiple_times(list_of_dicts, when)
    nodes = _merge_dictionaries(list_of_dicts)
    return nodes


@attr.s(frozen=True)
class _Placeholder:
    """A placeholder to mark unspecified keys in dictionaries."""

    scalar = attr.ib(type=bool, default=False)
    id_ = attr.ib(factory=uuid.uuid4, type=uuid.UUID)


def _convert_to_dict(x: Any, first_level: bool = True) -> Any | dict[Any, Any]:
    """Convert any object to a dictionary."""
    if isinstance(x, dict):
        return {k: _convert_to_dict(v, False) for k, v in x.items()}
    elif isinstance(x, Iterable) and not isinstance(x, str):
        if first_level:
            return {
                _Placeholder(): _convert_to_dict(element, False)
                for i, element in enumerate(x)
            }
        else:
            return {i: _convert_to_dict(element, False) for i, element in enumerate(x)}
    elif first_level:
        return {_Placeholder(scalar=True): x}
    else:
        return x


def _collect_node(
    session: Session, path: Path, name: str, node: str | Path
) -> dict[str, MetaNode]:
    """Collect nodes for a task.

    Parameters
    ----------
    session : _pytask.session.Session
        The session.
    path : Path
        The path to the task whose nodes are collected.
    name : str
        The name of the task.
    nodes : Dict[str, Union[str, Path]]
        A dictionary of nodes parsed from the ``depends_on`` or ``produces`` markers.

    Returns
    -------
    Dict[str, MetaNode]
        A dictionary of node names and their paths.

    Raises
    ------
    NodeNotCollectedError
        If the node could not collected.

    """
    collected_node = session.hook.pytask_collect_node(
        session=session, path=path, node=node
    )
    if collected_node is None:
        raise NodeNotCollectedError(
            f"{node!r} cannot be parsed as a dependency or product for task "
            f"{name!r} in {path!r}."
        )

    return collected_node


def _check_that_names_are_not_used_multiple_times(
    list_of_dicts: list[dict[Any, Any]], when: str
) -> None:
    """Check that names of nodes are not assigned multiple times.

    Tuples in the list have either one or two elements. The first element in the two
    element tuples is the name and cannot occur twice.

    """
    names_with_provisional_keys = list(
        itertools.chain.from_iterable(dict_.keys() for dict_ in list_of_dicts)
    )
    names = [x for x in names_with_provisional_keys if not isinstance(x, _Placeholder)]
    duplicated = find_duplicates(names)

    if duplicated:
        raise ValueError(
            f"'@pytask.mark.{when}' has nodes with the same name: {duplicated}"
        )


def _union_of_dictionaries(dicts: list[dict[Any, Any]]) -> dict[Any, Any]:
    """Merge multiple dictionaries in one.

    Examples
    --------
    >>> a, b = {"a": 0}, {"b": 1}
    >>> _union_of_dictionaries([a, b])
    {'a': 0, 'b': 1}

    >>> a, b = {'a': 0}, {'a': 1}
    >>> _union_of_dictionaries([a, b])
    {'a': 1}

    """
    return dict(itertools.chain.from_iterable(dict_.items() for dict_ in dicts))


def _merge_dictionaries(list_of_dicts: list[dict[Any, Any]]) -> dict[Any, Any]:
    """Merge multiple dictionaries.

    The function does not perform a deep merge. It simply merges the dictionary based on
    the first level keys which are either unique names or placeholders. During the merge
    placeholders will be replaced by an incrementing integer.

    Examples
    --------
    >>> a, b = {"a": 0}, {"b": 1}
    >>> _merge_dictionaries([a, b])
    {'a': 0, 'b': 1}

    >>> a, b = {_Placeholder(): 0}, {_Placeholder(): 1}
    >>> _merge_dictionaries([a, b])
    {0: 0, 1: 1}

    """
    merged_dict = _union_of_dictionaries(list_of_dicts)

    if len(merged_dict) == 1 and isinstance(list(merged_dict)[0], _Placeholder):
        placeholder, value = list(merged_dict.items())[0]
        if placeholder.scalar:
            out = value
        else:
            out = {0: value}
    else:
        counter = itertools.count()
        out = {}
        for k, v in merged_dict.items():
            if isinstance(k, _Placeholder):
                while True:
                    possible_key = next(counter)
                    if possible_key not in merged_dict and possible_key not in out:
                        out[possible_key] = v
                        break
            else:
                out[k] = v

    return out
