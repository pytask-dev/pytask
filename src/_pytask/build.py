"""Implement the build command."""
from __future__ import annotations

import sys
from typing import Any
from typing import TYPE_CHECKING

import click
from _pytask.click import ColoredCommand
from _pytask.config import hookimpl
from _pytask.console import console
from _pytask.exceptions import CollectionError
from _pytask.exceptions import ConfigurationError
from _pytask.exceptions import ExecutionError
from _pytask.exceptions import ResolvingDependenciesError
from _pytask.outcomes import ExitCode
from _pytask.pluginmanager import get_plugin_manager
from _pytask.session import Session
from _pytask.traceback import remove_internal_traceback_frames_from_exc_info
from rich.traceback import Traceback


if TYPE_CHECKING:
    from typing import NoReturn


@hookimpl(tryfirst=True)
def pytask_extend_command_line_interface(cli: click.Group) -> None:
    """Extend the command line interface."""
    cli.add_command(build)


def main(config_from_cli: dict[str, Any]) -> Session:
    """Run pytask.

    This is the main command to run pytask which usually receives kwargs from the
    command line interface. It can also be used to run pytask interactively. Pass
    configuration in a dictionary.

    Parameters
    ----------
    config_from_cli : dict[str, Any]
        A dictionary with options passed to pytask. In general, this dictionary holds
        the information passed via the command line interface.

    Returns
    -------
    session : _pytask.session.Session
        The session captures all the information of the current run.

    """
    try:
        pm = get_plugin_manager()
        from _pytask import cli

        pm.register(cli)
        pm.hook.pytask_add_hooks(pm=pm)

        config = pm.hook.pytask_configure(pm=pm, config_from_cli=config_from_cli)

        session = Session.from_config(config)

    except (ConfigurationError, Exception):
        exc_info = sys.exc_info()
        exc_info = remove_internal_traceback_frames_from_exc_info(exc_info)
        traceback = Traceback.from_exception(*exc_info)
        console.print(traceback)
        session = Session({}, None)
        session.exit_code = ExitCode.CONFIGURATION_FAILED

    else:
        try:
            session.hook.pytask_log_session_header(session=session)
            session.hook.pytask_collect(session=session)
            session.hook.pytask_resolve_dependencies(session=session)
            session.hook.pytask_execute(session=session)

        except CollectionError:
            session.exit_code = ExitCode.COLLECTION_FAILED

        except ResolvingDependenciesError:
            session.exit_code = ExitCode.RESOLVING_DEPENDENCIES_FAILED

        except ExecutionError:
            session.exit_code = ExitCode.FAILED

        except Exception:
            exc_info = sys.exc_info()
            exc_info = remove_internal_traceback_frames_from_exc_info(exc_info)
            traceback = Traceback.from_exception(*exc_info)
            console.print(traceback)
            session.exit_code = ExitCode.FAILED

        session.hook.pytask_unconfigure(session=session)

    return session


@click.command(cls=ColoredCommand)
@click.option(
    "--debug-pytask",
    is_flag=True,
    default=None,
    help="Trace all function calls in the plugin framework. [dim]\\[default: False][/]",
)
@click.option(
    "-x",
    "--stop-after-first-failure",
    is_flag=True,
    default=None,
    help="Stop after the first failure.",
)
@click.option("--max-failures", default=None, help="Stop after some failures.")
@click.option(
    "--show-errors-immediately",
    is_flag=True,
    default=None,
    help="Print errors with tracebacks as soon as the task fails.",
)
@click.option(
    "--show-traceback/--show-no-traceback",
    type=bool,
    default=None,
    help=(
        "Choose whether tracebacks should be displayed or not. "
        "[dim]\\[default: True][/]"
    ),
)
@click.option("--dry-run", type=bool, is_flag=True, help="Perform a dry-run.")
def build(**config_from_cli: Any) -> NoReturn:
    """Collect tasks, execute them and report the results.

    This is pytask's default command. pytask collects tasks from the given paths or the
    current working directory, executes them and reports the results.

    """
    config_from_cli["command"] = "build"
    session = main(config_from_cli)
    sys.exit(session.exit_code)
