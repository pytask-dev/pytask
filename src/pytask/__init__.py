"""This module contains the main namespace for pytask."""
from __future__ import annotations

from _pytask import __version__
from _pytask.build import main
from _pytask.collect_utils import depends_on
from _pytask.collect_utils import parse_nodes
from _pytask.collect_utils import produces
from _pytask.compat import check_for_optional_program
from _pytask.compat import import_optional_dependency
from _pytask.config import hookimpl
from _pytask.console import console
from _pytask.database import db
from _pytask.exceptions import CollectionError
from _pytask.exceptions import ConfigurationError
from _pytask.exceptions import ExecutionError
from _pytask.exceptions import NodeNotCollectedError
from _pytask.exceptions import NodeNotFoundError
from _pytask.exceptions import PytaskError
from _pytask.exceptions import ResolvingDependenciesError
from _pytask.graph import build_dag
from _pytask.mark import Mark
from _pytask.mark import MARK_GEN as mark  # noqa: N811
from _pytask.mark import MarkDecorator
from _pytask.mark import MarkGenerator
from _pytask.mark_utils import get_all_marks
from _pytask.mark_utils import get_marks
from _pytask.mark_utils import has_mark
from _pytask.mark_utils import remove_marks
from _pytask.mark_utils import set_marks
from _pytask.models import CollectionMetadata
from _pytask.nodes import FilePathNode
from _pytask.nodes import MetaNode
from _pytask.nodes import Task
from _pytask.outcomes import CollectionOutcome
from _pytask.outcomes import count_outcomes
from _pytask.outcomes import Exit
from _pytask.outcomes import ExitCode
from _pytask.outcomes import Persisted
from _pytask.outcomes import Skipped
from _pytask.outcomes import SkippedAncestorFailed
from _pytask.outcomes import SkippedUnchanged
from _pytask.outcomes import TaskOutcome
from _pytask.report import CollectionReport
from _pytask.report import ExecutionReport
from _pytask.report import ResolvingDependenciesReport
from _pytask.session import Session
from _pytask.traceback import format_exception_without_traceback
from _pytask.traceback import remove_internal_traceback_frames_from_exc_info
from _pytask.traceback import remove_traceback_from_exc_info
from _pytask.traceback import render_exc_info
from _pytask.warnings_utils import parse_warning_filter
from _pytask.warnings_utils import warning_record_to_str
from _pytask.warnings_utils import WarningReport


# This import must come last, otherwise a circular import occurs.
from _pytask.cli import cli  # noreorder


__all__ = [
    "__version__",
    "build_dag",
    "check_for_optional_program",
    "cli",
    "console",
    "count_outcomes",
    "db",
    "depends_on",
    "format_exception_without_traceback",
    "get_all_marks",
    "get_first_non_none_value",
    "get_marks",
    "has_mark",
    "hookimpl",
    "import_optional_dependency",
    "main",
    "mark",
    "parse_nodes",
    "parse_warning_filter",
    "produces",
    "remove_internal_traceback_frames_from_exc_info",
    "remove_marks",
    "remove_traceback_from_exc_info",
    "render_exc_info",
    "set_marks",
    "warning_record_to_str",
    "CollectionError",
    "CollectionMetadata",
    "CollectionOutcome",
    "CollectionReport",
    "ConfigurationError",
    "ExecutionReport",
    "ExecutionError",
    "Exit",
    "ExitCode",
    "FilePathNode",
    "Mark",
    "MarkDecorator",
    "MarkGenerator",
    "MetaNode",
    "NodeNotCollectedError",
    "NodeNotFoundError",
    "Persisted",
    "PytaskError",
    "ResolvingDependenciesError",
    "ResolvingDependenciesReport",
    "Session",
    "Skipped",
    "SkippedAncestorFailed",
    "SkippedUnchanged",
    "Task",
    "TaskOutcome",
    "WarningReport",
]
