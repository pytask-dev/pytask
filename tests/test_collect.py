import os
import textwrap

import pytest
from _pytask.exceptions import NodeNotCollectedError
from pytask import main


@pytest.mark.end_to_end
def test_collect_filepathnode_with_relative_path(tmp_path):
    source = """
    import pytask
    from pathlib import Path


    @pytask.mark.depends_on("in.txt")
    @pytask.mark.produces("out.txt")
    def task_dummy():
        Path("out.txt").write_text(
            Path("in.txt").read_text()
        )
    """
    tmp_path.joinpath("task_dummy.py").write_text(textwrap.dedent(source))
    tmp_path.joinpath("in.txt").write_text("Relative paths work.")

    os.chdir(tmp_path)
    session = main({"paths": tmp_path})

    assert session.collection_reports[0].successful
    assert tmp_path.joinpath("out.txt").read_text() == "Relative paths work."


@pytest.mark.end_to_end
def test_collect_filepathnode_with_unknown_type(tmp_path):
    """If a node cannot be parsed because unknown type, raise an error."""
    source = """
    import pytask

    @pytask.mark.depends_on(True)
    def task_with_non_path_dependency():
        pass
    """
    tmp_path.joinpath("task_dummy.py").write_text(textwrap.dedent(source))

    os.chdir(tmp_path)
    session = main({"paths": tmp_path})

    assert session.exit_code == 2
    assert isinstance(session.collection_reports[0].exc_info[1], NodeNotCollectedError)


@pytest.mark.end_to_end
def test_collect_nodes_with_the_same_name(tmp_path):
    """Nodes with the same filename, not path, are not mistaken for each other."""
    source = """
    import pytask
    from pathlib import Path

    @pytask.mark.depends_on("text.txt")
    @pytask.mark.produces("out_0.txt")
    def task_0():
        in_ = Path("text.txt").read_text()
        print(Path("text.txt").resolve())
        Path("out_0.txt").write_text(in_)

    @pytask.mark.depends_on("sub/text.txt")
    @pytask.mark.produces("out_1.txt")
    def task_1():
        in_ = Path("sub/text.txt").read_text()
        Path("out_1.txt").write_text(in_)
    """
    tmp_path.joinpath("task_dummy.py").write_text(textwrap.dedent(source))

    tmp_path.joinpath("text.txt").write_text("in root")

    tmp_path.joinpath("sub").mkdir()
    tmp_path.joinpath("sub", "text.txt").write_text("in sub")

    os.chdir(tmp_path)
    session = main({"paths": tmp_path})

    assert session.exit_code == 0
    assert tmp_path.joinpath("out_0.txt").read_text() == "in root"
    assert tmp_path.joinpath("out_1.txt").read_text() == "in sub"
