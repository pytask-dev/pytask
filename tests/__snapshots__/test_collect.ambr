# serializer version: 1
# name: test_error_when_dependency_is_defined_in_kwargs_and_annotation
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 0 tasks.
  
  ────────────────────────── Failures during collection ──────────────────────────
  ─ Could not collect /tmp/pytest-of-tobia/pytest-1709/test_error_when_dependen… ─
  
  ValueError: The value for the parameter 'in_' is defined twice in 
  '@pytask.mark.task(kwargs=...)' and in the type annotation. Choose only one 
  option.
  
  ╭───────────────── Summary ─────────────────╮
  │  1  Collected errors and tasks            │
  │  1  Failed                      (100.0%)  │
  ╰───────────────────────────────────────────╯
  ──────────────────────────── Failed in 0.0 seconds ─────────────────────────────
  '''
# ---
